### PR TITLE
[WIP] blackbox original sources

### LINKED ISSUE #7212: [blackbox]  unable to blackbox original sources
User Agent: Mozilla/5.0 (Macintosh; Intel Mac OS X 10.13; rv:62.0) Gecko/20100101 Firefox/62.0
Build ID: 20180830143136

Steps to reproduce:

1. Create webpack site with multiple js sources
2. Include the bundle in an html page, ensure the js is executed
3. Run the site using webpack dev server.
4. Open devtools, debugger tab
5. Expand Webpack view, and path to target source file
6. Click the 'eye' icon
alternatively:
6. Right click on the source

Actual results:

When clicking the eye icon, nothing happens.
When right clicking a context menu appears. "Blacklist source" is grayed out.

Expected results:

Source should have been added to blacklist, indicator present on icon. If that's not possible then an error message should have appeared explaining why; and there should be documentation about the limitations of blacklist source.

### PR DESCRIPTION
**Pull request by AnshulMalik**
Fixes #7212 

### Summary of Changes

* Allow original sources to be blackboxed

### Test Plan

Still need to write tests

### Screenshots/Videos (OPTIONAL)
optimize

### CODE REVIEW

File: src/actions/sources/blackbox.js [hunk 1]
```diff
@@ -17,16 +17,45 @@ import type { ThunkArgs } from "../types";
...
+      source
+    );
+
+    const lines = (await sourceMaps.getOriginalSourceText(source)).text.split(
+      "\n"
+    );
+
+    let endLocation = { line: null };
+    let attempt = 0;
+    while (endLocation.line === null) {
```
Comment (AnshulMalik) [inline src/actions/sources/blackbox.js]: this is the hacky version of finding the end, let me know if anyone knows better way to do it.

File: src/actions/sources/blackbox.js [hunk 2]
```diff
@@ -17,16 +17,45 @@ import type { ThunkArgs } from "../types";
...
+      const line = lines.length - attempt;
+      const column = lines[line - 1].length - 1;
```
Comment (jasonLaster) [inline src/actions/sources/blackbox.js]: lets move this to a utility and perhaps double check that stack overflow doesn't have something better

File: src/actions/sources/blackbox.js [hunk 3]
```diff
@@ -17,16 +17,45 @@ import type { ThunkArgs } from "../types";
...
+      console.log(line, column);
+      endLocation = await sourceMaps.getGeneratedLocation(
+        {
+          sourceId: source.id,
+          line,
+          column: column < 0 ? 0 : column
```
Comment (jasonLaster) [inline src/actions/sources/blackbox.js]: Math.min

File: src/actions/sources/blackbox.js [hunk 4]
```diff
@@ -17,16 +17,45 @@ import type { ThunkArgs } from "../types";
...
+      "\n"
+    );
+
+    let endLocation = { line: null };
+    let attempt = 0;
+    while (endLocation.line === null) {
+      const line = lines.length - attempt;
+      const column = lines[line - 1].length - 1;
+      console.log(line, column);
+      endLocation = await sourceMaps.getGeneratedLocation(
```
Comment (jasonLaster) [inline src/actions/sources/blackbox.js]: why not map at the end... as in find the last line/col and map that. Lets assume a good world :)

File: src/actions/sources/blackbox.js [hunk 5]
```diff
@@ -17,16 +17,45 @@ import type { ThunkArgs } from "../types";
...
+    }
+
     return dispatch({
       type: "BLACKBOX",
       source,
-      [PROMISE]: client.blackBox(id, isBlackBoxed)
+      [PROMISE]: client.blackBox(startLocation.sourceId, isBlackBoxed, {
+        start: startLocation,
+        end: endLocation
+      })
```
Comment (jasonLaster) [inline src/actions/sources/blackbox.js]: wow - we dont have to update the reducer at all.  Thats pretty cool

File: src/actions/sources/blackbox.js (hunk 2 above)
Comment (loganfsmyth) [reply src/actions/sources/blackbox.js]: I think my vote would be for this to live in `devtools-source-map`. Maybe `sourceMaps.getFileGeneratedRange(source)` or something? Then we have full access to the whole map easily.

File: src/actions/sources/blackbox.js (hunk 1 above)
Comment (loganfsmyth) [reply src/actions/sources/blackbox.js]: If this is moved to be part of the worker, we can use the SourceMapConsumer's `generatedPositionFor` with `bias: GREATEST_LOWER_BOUND` to do this automatic backward slide.

File: src/actions/sources/blackbox.js (hunk 2 above)
Comment (jasonLaster) [reply src/actions/sources/blackbox.js]: that makes a lot of sense

File: src/actions/sources/blackbox.js (hunk 4 above)
Comment (AnshulMalik) [reply src/actions/sources/blackbox.js]: I tried mapping the last line and it's last column, but it mapped to nothing. Then I tried to go up the lines to eventually find a mapping. 
This looks bad to me, still trying to find a better way

File: src/actions/sources/blackbox.js (hunk 2 above)
Comment (AnshulMalik) [reply src/actions/sources/blackbox.js]: Yeah, that's better

File: src/actions/sources/blackbox.js (hunk 1 above)
Comment (AnshulMalik) [reply src/actions/sources/blackbox.js]: awesome, will try that

File: src/actions/sources/blackbox.js (hunk 4 above)
Comment (jasonLaster) [reply src/actions/sources/blackbox.js]: CC @loganfsmyth mind sketching a proof of concept

File: packages/devtools-source-map/src/source-map.js [hunk 6]
```diff
@@ -306,6 +306,45 @@ async function getOriginalSourceText(originalSource: Source) {
...
+  }
+
+  const start = await getGeneratedLocation(
+    { sourceId: originalSource.id, line: 1, column: 0 },
+    originalSource
+  );
+
+  const lines = sourceText.split("\n");
+
+  const endRanges = await getGeneratedRanges(
```
Comment (jasonLaster) [inline packages/devtools-source-map/src/source-map.js]: i spoke to @loganfsmyth and we can drop this approach in favor of something much simpler...

for the end location we should get the generatedPosition for the end line/column with a SourceMapConsumer.LEAST_UPPER_BOUND bias. This should just be one call..

File: packages/devtools-source-map/src/utils/sourceMapRequests.js [hunk 7]
```diff
@@ -16,8 +16,20 @@ function setSourceMap(generatedId, request) {
...
+    return;
+  }
+
+  return await request;
+}
+
 module.exports = {
   clearSourceMaps,
   getSourceMap,
+  getSourceMapUnrwapped,
```
[the remark this answers is not in the document]
Comment (AnshulMalik) [reply packages/devtools-source-map/src/utils/sourceMapRequests.js]: oops, forgot to remove

**Pull request merged**